data/src/main/java/co/nimblehq/compose/crypto/data/repository/CoinRepositoryImpl.kt

```diff
@@ -0,0 +1,35 @@
+package co.nimblehq.compose.crypto.data.repository
+
+import co.nimblehq.compose.crypto.data.model.response.toModels
+import co.nimblehq.compose.crypto.data.service.ApiService
+import co.nimblehq.compose.crypto.data.transform
+import co.nimblehq.compose.crypto.domain.model.CoinItem
+import co.nimblehq.compose.crypto.domain.repository.CoinRepository
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.map
+
+class CoinRepositoryImpl(
+    private val api: ApiService
+) : CoinRepository {
+
+    override fun getTrendingList(
+        coinIds: List<String>,
+        currency: String,
+        priceChangePercentage: String,
+        itemOrder: String,
+        itemPerPage: Int,
+        page: Int
+    ): Flow<List<CoinItem>> = flow {
+        emit(
+            api.getCoins(
+                coinIds = coinIds.joinToString(separator = ","),
+                currency = currency,
+                priceChangePercentage = priceChangePercentage,
+                itemOrder = itemOrder,
+                itemPerPage = itemPerPage,
+                page = page
+            )
+        )
+    }.transform().map { it.toModels() }
```

Comment:
@hoangnguyen92dn I tried to test the `usecase` and `repository`, but I didn't get any `result` value. Then I tried to use `try/catch`, and I found the error. It seems like we are facing the issue of parsing the type of `Long` and `Int` values from `Moshi` and here's some error:
<img width="851" alt="Screen Shot 2565-08-05 at 10 47 50" src="https://user-images.githubusercontent.com/28002315/183002924-e883162c-39d8-4108-aba7-55b53e2e98a8.png">

By the way, I tried to use `Double` and it works but your current implementation it still doesn't work, I don't get any result, Correct me if I missed something 🙇‍♀️

Reply:
@Wadeewee You're right. I am using the wrong data type. I will update krub 🚀

Reply:
We can upgrade the transform extension to catch the http ex @hoangnguyen92dn. This is what I applied in Worpt project

```
fun <T> flowErrorMapping(@BuilderInference block: suspend FlowCollector<T>.() -> T) = flow {
    val result = try {
        block()
    } catch (e: Exception) {
        throw e.mapError()
    }
    emit(result)
}

@Suppress("NestedBlockDepth, ComplexMethod")
private fun Throwable.mapError(): Throwable {
    return when {
        this is UnknownHostException || this is InterruptedIOException -> NoConnectivityException
        this is HttpException -> {
            val errorResponse = getErrorBodyObject<ErrorResponse>()
            return when (errorResponse?.firstErrorCode) {
                ERROR_CODE_VALIDATION_ERROR,
                ERROR_CODE_UNPROCESSABLE_ENTITY -> ValidationException(errorResponse)
                ERROR_CODE_UNAUTHORIZED -> UnauthorizedException(errorResponse)
                ERROR_VERIFY_USER_BALANCE -> InsufficientBalanceException(errorResponse)
                else -> GenericException(errorResponse)
            }
        }
        isCanceledException(this) -> CanceledException
        else -> this
    }
}
```

Then this is what we can do in Repository layer

```
    override fun createUser(
        firstName: String,
        lastName: String,
        email: String,
        password: String,
        termsAndConditionId: String,
        privacyPolicyId: String
    ): Flow<AuthorizationCode> = flowErrorMapping {
        clientAuthenticationApiService.createUser(
            CreateUserRequest(
                firstName = firstName,
                lastName = lastName,
                email = email,
                password = password,
                termsAndConditionId = termsAndConditionId,
                privacyPolicyId = privacyPolicyId
            )
        ).toAuthorizationCode()
    }
```